crates/bevy_render/src/render_resource/storage_buffer.rs

```diff
@@ -1,141 +1,153 @@
-                size = (size + <U as AsStd430>::Output::ALIGNMENT - 1)
-                    & !(<U as AsStd430>::Output::ALIGNMENT - 1);
-            }
-            // Variable size arrays must have at least 1 element
-            size += Self::ITEM_SIZE * self.values.len().max(1);
+impl<T: ShaderType> Default for DynamicStorageBuffer<T> {
+    fn default() -> Self {
+        Self {
+            values: Vec::new(),
+            scratch: DynamicStorageBufferWrapper::new(Vec::new()),
+            buffer: None,
+            capacity: 0,
         }
-        size
     }
+}
 
-    pub fn write_buffer(&mut self, device: &RenderDevice, queue: &RenderQueue) {
-        self.reserve_buffer(device);
-        if let Some(storage_buffer) = &self.storage_buffer {
-            let range = 0..self.size();
-            let mut writer = std430::Writer::new(&mut self.scratch[range.clone()]);
-            let mut offset = 0;
-            // First write the struct body if there is one
-            if Self::BODY_SIZE > 0 {
-                if let Ok(new_offset) = writer.write(&self.body).map_err(|e| warn!("{:?}", e)) {
-                    offset = new_offset;
-                }
-            }
-            if Self::ITEM_SIZE > 0 {
-                if self.values.is_empty() {
-                    // Zero-out the padding and dummy array item in the case of the array being empty
-                    for i in offset..self.size() {
-                        self.scratch[i] = 0;
-                    }
-                } else {
-                    // Then write the array. Note that padding bytes may be added between the body
-                    // and the array in order to align the array to the alignment requirements of its
-                    // items
-                    writer
-                        .write(self.values.as_slice())
-                        .map_err(|e| warn!("{:?}", e))
-                        .ok();
-                }
-            }
-            queue.write_buffer(storage_buffer, 0, &self.scratch[range]);
-        }
+impl<T: ShaderType + WriteInto> DynamicStorageBuffer<T> {
+    #[inline]
+    pub fn buffer(&self) -> Option<&Buffer> {
+        self.buffer.as_ref()
     }
 
-    pub fn values(&self) -> &[T] {
-        &self.values
+    #[inline]
+    pub fn binding(&self) -> Option<BindingResource> {
+        Some(BindingResource::Buffer(BufferBinding {
+            buffer: self.buffer()?,
+            offset: 0,
+            size: Some(T::min_size()),
+        }))
     }
 
-    pub fn values_mut(&mut self) -> &mut [T] {
-        &mut self.values
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.values.len()
     }
 
     #[inline]
-    pub fn clear(&mut self) {
-        self.values.clear();
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
     }
 
     #[inline]
-    pub fn push(&mut self, value: T) {
+    pub fn push(&mut self, value: T) -> u32 {
+        let offset = self.scratch.write(&value).unwrap() as u32;
```

Comment:
Does this mean that inter-item padding for alignment happens when calling `.write()`? That would make sense. Just noting it for myself.